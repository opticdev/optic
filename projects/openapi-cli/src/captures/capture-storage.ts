import Path from 'path';
import * as fs from 'fs-extra';
import { InputErrors } from '../commands/reporters/feedback';
import os from 'os';
const tmpDirectory = os.tmpdir();
import crypto from 'crypto';

export async function captureStorage(
  filePath: string
): Promise<[boolean, string]> {
  const resolvedFilepath = Path.resolve(filePath);
  const openApiExists: boolean = await fs.pathExists(resolvedFilepath);

  const specPathHash = crypto
    .createHash('md5')
    .update(resolvedFilepath)
    .digest('hex');

  const trafficDirectory = Path.join(
    tmpDirectory,
    'optic',
    'captures',
    specPathHash
  );

  if (openApiExists) await fs.ensureDir(trafficDirectory);

  return [openApiExists, trafficDirectory];
}
