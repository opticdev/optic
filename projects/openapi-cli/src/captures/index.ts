import { CapturedBody } from './body';
import { CapturedBodies } from './streams/captured-bodies';

export { CapturedBody, CapturedBodies };
export { CapturedInteractions } from './streams/captured-interactions';

export { CapturedInteraction } from './interaction';
export type { CapturedRequest, CapturedResponse } from './interaction';

export { HarEntries } from './streams/sources/har';
export type { HttpArchive } from './streams/sources/har';

export { ProxyInteractions } from './streams/sources/proxy';
