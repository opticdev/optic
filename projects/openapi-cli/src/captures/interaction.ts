import { CapturedBody } from './body';
import { OpenAPIV3 } from '../specs';
import { HttpArchive } from './streams/sources/har';
import { ProxySource } from './streams/sources/proxy';
import { URL } from 'url';
import { HttpMethods, Operation } from '../operations';
import invariant from 'ts-invariant';
import { Buffer } from 'buffer';

export interface CapturedInteraction {
  request: {
    host: string;
    method: OpenAPIV3.HttpMethods;
    path: string;
    body: CapturedBody | null;
    // TODO: add support for headers and query params
  };
  response: {
    statusCode: string;
    body: CapturedBody | null;
    // TODO: add support headers
  };
}
export class CapturedInteraction {
  static fromHarEntry(entry: HttpArchive.Entry): CapturedInteraction {
    const url = new URL(entry.request.url);

    const method = HttpMethods[entry.request.method];
    invariant(
      Operation.isHttpMethod(method),
      `expect HAR entry to have a valid request method`
    );

    let requestBody: CapturedBody | null = null;
    let responseBody: CapturedBody | null = null;

    const requestPostData = entry.request.postData;
    if (
      requestPostData &&
      (!requestPostData.encoding || Buffer.isEncoding(requestPostData.encoding))
    ) {
      let buffer = Buffer.from(
        requestPostData.text,
        requestPostData.encoding as BufferEncoding | undefined
      );
      requestBody = CapturedBody.from(
        buffer,
        requestPostData.mimeType,
        entry.request.bodySize
      );
    }

    const responseContent = entry.response.content;
    if (
      responseContent.text &&
      (!responseContent.encoding || Buffer.isEncoding(responseContent.encoding))
    ) {
      let buffer = Buffer.from(
        responseContent.text,
        responseContent.encoding as BufferEncoding | undefined
      );
      responseBody = CapturedBody.from(
        buffer,
        responseContent.mimeType,
        responseContent.size
      );
    }

    return {
      request: {
        host: url.hostname,
        method,
        path: url.pathname,
        body: requestBody,
      },
      response: {
        statusCode: '' + entry.response.status,
        body: responseBody,
      },
    };
  }

  static fromProxyInteraction(
    proxyInteraction: ProxySource.Interaction
  ): CapturedInteraction {
    const url = new URL(proxyInteraction.request.url);

    const method = HttpMethods[proxyInteraction.request.method];
    invariant(
      Operation.isHttpMethod(method),
      `expect proxy interaction to have a valid request method`
    );

    let requestBody: CapturedBody | null = null;
    let responseBody: CapturedBody | null = null;

    const requestBodyBuffer = proxyInteraction.request.body.buffer;
    if (requestBodyBuffer.length > 0) {
      let contentType = proxyInteraction.request.headers['content-type'];
      let contentLength = proxyInteraction.request.headers['content-length'];

      requestBody = CapturedBody.from(
        requestBodyBuffer,
        contentType || null,
        contentLength ? parseInt(contentLength, 10) : 0
      );
    }

    const responseBodyBuffer = proxyInteraction.response.body.buffer;
    if (responseBodyBuffer.length > 0) {
      let contentType = proxyInteraction.response.headers['content-type'];
      let contentLength = proxyInteraction.response.headers['content-length'];

      responseBody = CapturedBody.from(
        responseBodyBuffer,
        contentType || null,
        contentLength ? parseInt(contentLength, 10) : 0
      );
    }

    return {
      request: {
        host: url.hostname,
        method,
        path: url.pathname,
        body: requestBody,
      },
      response: {
        statusCode: '' + proxyInteraction.response.statusCode,
        body: responseBody,
      },
    };
  }
}

export type CapturedRequest = CapturedInteraction['request'];
export type CapturedResponse = CapturedInteraction['response'];
