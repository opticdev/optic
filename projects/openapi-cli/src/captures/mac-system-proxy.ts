export function parseMacNetworkLine(line: string): {
  enabled: boolean;
  host: string;
  port: string;
  authenticated: boolean;
} {
  const match = line.match(
    /Enabled: (Yes|No)\nServer: ([:/\da-zA-Z.\-_]*)\nPort: (\d+)\nAuthenticated Proxy Enabled: (0|1)/
  );

  if (match) {
    const [_, enabled, host, port, authenticated] = match;

    return {
      enabled: enabled === 'Yes',
      host,
      port,
      authenticated: authenticated === 'Yes',
    };
  } else {
    throw new Error(`Unexpected proxy config returned from networksetup`);
  }
}
