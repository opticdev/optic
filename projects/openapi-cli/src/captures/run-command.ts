import { ChildProcessWithoutNullStreams, spawn } from 'child_process';
import { createCommandFeedback } from '../commands/reporters/feedback';
import url from 'url';

export class RunCommand {
  constructor(
    private proxyUrl: string,
    private feedback: ReturnType<typeof createCommandFeedback>
  ) {}

  completed: boolean = false;
  process: ChildProcessWithoutNullStreams | undefined = undefined;

  async run(command: string): Promise<{ exitCode: number }> {
    const { port, hostname } = url.parse(this.proxyUrl);

    const [commandStart] = command.split(/ +/);

    const commandProcess = spawn(command, {
      cwd: process.cwd(),
      env: {
        ...process.env,
        http_proxy: `http://${hostname}:${port}`,
        https_proxy: `https://${hostname}:${port}`,
      },
      shell: true,
    });

    this.feedback.notable('Running command: ' + command);

    this.process = commandProcess;

    commandProcess.stdout.setEncoding('utf8');

    commandProcess.stdout.on('data', (data) => {
      const str = data.toString(),
        lines = str.split(/(\r?\n)/g);
      lines.forEach(
        (line) => line.trim() && this.feedback.logChild(commandStart, line)
      );
    });

    commandProcess.stderr.on('data', (data) => {
      const str = data.toString(),
        lines = str.split(/(\r?\n)/g);
      lines.forEach(
        (line) => line.trim() && this.feedback.logChild(commandStart, line)
      );
    });

    return new Promise((resolve) => {
      const endIt = (code: number | undefined) => {
        this.completed = true;
        resolve({ exitCode: commandProcess?.exitCode || 0 });
        commandProcess!.removeAllListeners('error');
        commandProcess!.removeAllListeners('data');
        commandProcess!.removeAllListeners('exit');
      };
      commandProcess!.once('exit', (code) => endIt(code || undefined));
      commandProcess!.once('end', endIt);
      commandProcess!.once('close', endIt);
    });
  }

  async kill() {
    if (!this.completed && this.process) {
      this.process.kill(1);
    }
  }
}
