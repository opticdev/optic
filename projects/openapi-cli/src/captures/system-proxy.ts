import url from 'url';
import { createCommandFeedback } from '../commands/reporters/feedback';
import { platform, runCommand } from '../shell-utils';

export class SystemProxy {
  constructor(
    private proxyUrl: string,
    private feedback: ReturnType<typeof createCommandFeedback>
  ) {}
  private stopCommand: (() => Promise<void>) | undefined = undefined;

  async start(device: string | undefined) {
    const { port } = url.parse(this.proxyUrl);

    if (platform === 'mac') {
      const name = await chooseInterfaceMac();

      await Promise.all([
        runCommand(`networksetup -setwebproxy "${name}" 127.0.0.1 ${port}`),
        runCommand(
          `networksetup -setsecurewebproxy "${name}" 127.0.0.1 ${port}`
        ),
        runCommand(
          `networksetup -setproxybypassdomains "${name}" "<-loopback>"`
        ),
      ]);

      this.feedback.notable(
        `Proxy running on ${this.proxyUrl}. System proxy updated`
      );

      this.stopCommand = async () => {
        await Promise.all([
          runCommand(`networksetup -setwebproxystate "${name}" off`),
          runCommand(`networksetup -setsecurewebproxystate "${name}" off`),
        ]);
        this.feedback.notable(`Mac System Proxy settings cleared`);
      };
    } else {
      this.feedback.notable(
        `Proxy running on ${this.proxyUrl}. System proxy updated`
      );
      console.log(
        `automatic proxy configuration is not supported on ${platform}`
      );
    }
  }
  async stop() {
    if (this.stopCommand) await this.stopCommand();
  }
}

export async function chooseInterfaceMac() {
  const network = await runCommand('scutil --nwi');
  const device = new RegExp(/Network interfaces: (en[0-9])/);

  const result = network.match(device);

  let devicePreference = result ? result[1] : 'en0';

  const hardware = await runCommand('networksetup -listallhardwareports');
  const regex = new RegExp(/(Hardware Port: (.*)\nDevice: (.*)\n)/);

  let name: string | undefined = undefined;
  hardware.split(/\n\s*\n/).forEach((entry) => {
    if (name) return;
    const match = regex.exec(entry);

    if (match && match[3] === devicePreference) {
      name = match[2];
    }
  });
  return name || 'Wi-Fi';
}
