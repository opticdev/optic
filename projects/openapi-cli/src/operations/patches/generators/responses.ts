import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { OperationPatch } from '..';
import { Operation } from '../..';
import { PatchOperationGroup, PatchImpact } from '../../../patches';
import { OperationDiffResult, OperationDiffResultKind } from '../../diffs';
import { OpenAPIV3 } from '../../../specs';

export function* responsesPatches(
  diff: OperationDiffResult,
  operation: Operation,
  context?: {}
): IterableIterator<OperationPatch> {
  if (diff.kind === OperationDiffResultKind.UnmatchedResponseStatusCode) {
    yield* unmatchedStatusCode(diff, operation);
  } else if (diff.kind === OperationDiffResultKind.UnmatchedResponseBody) {
    yield* unmatchedResponseBody(diff, operation);
  } else if (diff.kind === OperationDiffResultKind.MissingResponseBody) {
    yield* missingResponseBody(diff, operation);
  }
}

function* unmatchedStatusCode(
  diff: OperationDiffResult & {
    kind: OperationDiffResultKind.UnmatchedResponseStatusCode;
  },
  _operation: Operation
): IterableIterator<OperationPatch> {
  let groupedOperations: PatchOperationGroup[] = [];

  const { statusCode, contentType } = diff;

  const numericalStatusCode = parseInt(statusCode, 10);
  if (
    numericalStatusCode < 200 ||
    numericalStatusCode >= 500 ||
    (numericalStatusCode >= 300 && numericalStatusCode < 400)
  ) {
    return; // only document 2xx and 4xx
  }

  const responseObject: OpenAPIV3.ResponseObject = {
    description: `${statusCode} response`, // required, no longer in v3.1
  };
  groupedOperations.push(
    PatchOperationGroup.create('add response status code', {
      op: 'add',
      path: jsonPointerHelpers.compile(['responses', statusCode]),
      value: responseObject,
    })
  );

  const mediaTypeObject: OpenAPIV3.MediaTypeObject = {};
  if (contentType) {
    groupedOperations.push(
      PatchOperationGroup.create(
        `add response body for content type '${contentType}'`,
        {
          op: 'add',
          path: jsonPointerHelpers.compile([
            'responses',
            statusCode,
            'content',
          ]),
          value: {
            [contentType]: mediaTypeObject,
          },
        }
      )
    );
  }

  yield {
    description: `add ${statusCode} response`,
    impact: [PatchImpact.Addition, PatchImpact.BackwardsCompatible],
    diff,
    groupedOperations,
  };
}

function* unmatchedResponseBody(
  diff: OperationDiffResult & {
    kind: OperationDiffResultKind.UnmatchedResponseBody;
  },
  operation: Operation
): IterableIterator<OperationPatch> {
  const { statusCode, contentType } = diff;
  const response = operation.responses[statusCode] as OpenAPIV3.ResponseObject;

  if (!contentType) return; // we can't add undefined content types

  const groupedOperations: PatchOperationGroup[] = [];

  if (!response.content) {
    groupedOperations.push(
      PatchOperationGroup.create(
        `add content object for ${statusCode} response`,
        {
          op: 'add',
          path: jsonPointerHelpers.compile([
            'responses',
            statusCode,
            'content',
          ]),
          value: {},
        }
      )
    );
  }

  const mediaTypeObject: OpenAPIV3.MediaTypeObject = {};
  groupedOperations.push(
    PatchOperationGroup.create(
      `add response body for content type '${contentType}'`,
      {
        op: 'add',
        path: jsonPointerHelpers.compile([
          'responses',
          statusCode,
          'content',
          contentType,
        ]),
        value: mediaTypeObject,
      }
    )
  );
  yield {
    description: `add ${contentType} response for ${statusCode}`,
    impact: [PatchImpact.Addition, PatchImpact.BackwardsCompatible],
    diff,
    groupedOperations,
  };
}

function* missingResponseBody(
  diff: OperationDiffResult & {
    kind: OperationDiffResultKind.MissingResponseBody;
  },
  operation: Operation
): IterableIterator<OperationPatch> {
  const { statusCode } = diff;
  const response = operation.responses[statusCode] as OpenAPIV3.ResponseObject;

  const contentSize = [...Object.keys(response.content || {})].length;

  const operationGroup = PatchOperationGroup.create(
    `remove response content (${
      contentSize > 0 ? `${contentSize} bodies` : 'empty'
    })`,
    {
      op: 'remove',
      path: jsonPointerHelpers.compile(['responses', statusCode, 'content']),
    }
  );

  yield {
    description: `remove response content for ${statusCode}`,
    impact: [
      contentSize > 0
        ? PatchImpact.BackwardsIncompatible
        : PatchImpact.BackwardsCompatible,
    ],
    diff,
    groupedOperations: [operationGroup],
  };
}
