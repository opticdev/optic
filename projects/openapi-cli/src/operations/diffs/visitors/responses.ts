import { CapturedResponse } from '../../../captures';
import { OpenAPIV3 } from '../../../specs';
import { OperationDiffResult, OperationDiffResultKind } from '../result';
import { findResponse, findBody } from '../..';

export function* visitResponses(
  capturedResponse: CapturedResponse,
  responses: { [statusCode: string]: OpenAPIV3.ResponseObject }
): IterableIterator<OperationDiffResult> {
  const responseMatch = findResponse(
    { responses },
    capturedResponse.statusCode
  );

  if (!responseMatch) {
    const statusCode = parseInt(capturedResponse.statusCode, 10);
    yield {
      kind: OperationDiffResultKind.UnmatchedResponseStatusCode,
      statusCode: capturedResponse.statusCode,
      contentType: capturedResponse.body?.contentType || null,
    };

    return; // no response, no more to diff
  }

  const [response] = responseMatch;

  const matchedBody = findBody(response, capturedResponse.body?.contentType);

  if (!matchedBody && capturedResponse.body) {
    yield {
      kind: OperationDiffResultKind.UnmatchedResponseBody,
      contentType: capturedResponse.body.contentType,
      statusCode: capturedResponse.statusCode,
    };
  } else if (response.content && !capturedResponse.body) {
    yield {
      kind: OperationDiffResultKind.MissingResponseBody,
      statusCode: capturedResponse.statusCode,
    };
  }
}
