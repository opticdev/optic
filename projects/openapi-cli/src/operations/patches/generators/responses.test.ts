import { responsesPatches } from './responses';
import {
  diffInteractionByOperation,
  OperationDiffResultKind,
} from '../../diffs';
import { OpenAPIV3 } from '../../../specs';
import { Operation, HttpMethods } from '../../';
import { CapturedInteraction, CapturedBody } from '../../../captures';
import { OperationPatch } from '..';

describe('responsesPatches', () => {
  it('generates a patch for an unspecified response status code', () => {
    const operation = operationFixture({
      '200': {
        description: 'success',
      },
    });

    const unspecifiedResponse = interactionFixture('400', null);

    let [diff] = [
      ...diffInteractionByOperation(unspecifiedResponse, operation),
    ];
    expect(diff.kind).toBe(OperationDiffResultKind.UnmatchedResponseStatusCode);

    let patches = [...responsesPatches(diff, operation)];
    expect(patches).toHaveLength(1);
    expect(patches).toMatchSnapshot();

    for (let patch of patches) {
      let patchedOperation = OperationPatch.applyTo(patch, operation).expect(
        'operation patch should apply to operation'
      );

      expect(patchedOperation).toMatchSnapshot();
      expect([
        ...diffInteractionByOperation(unspecifiedResponse, patchedOperation),
      ]).toHaveLength(0);
    }
  });

  it('does not generate a patch for an unspecified response status code other than 2xx, 3xx or 4xx', () => {
    const operation = operationFixture({
      '200': {
        description: 'success',
      },
    });

    const unspecifiedResponse = interactionFixture('500', null);

    let [diff] = [
      ...diffInteractionByOperation(unspecifiedResponse, operation),
    ];
    expect(diff.kind).toBe(OperationDiffResultKind.UnmatchedResponseStatusCode);

    let patches = [...responsesPatches(diff, operation)];
    expect(patches).toHaveLength(0);
  });

  it('generates a patch for an unspecified response body', () => {
    const operation = operationFixture({
      '200': {
        description: 'success',
        content: {
          'application/json': {},
        },
      },
    });

    const unspecifiedResponse = interactionFixture(
      '200',
      CapturedBody.from('test-content', 'text/plain')
    );

    let [diff] = [
      ...diffInteractionByOperation(unspecifiedResponse, operation),
    ];
    expect(diff.kind).toBe(OperationDiffResultKind.UnmatchedResponseBody);

    let patches = [...responsesPatches(diff, operation)];
    expect(patches).toHaveLength(1);
    expect(patches).toMatchSnapshot();

    for (let patch of patches) {
      let patchedOperation = OperationPatch.applyTo(patch, operation).expect(
        'operation patch should apply to operation'
      );

      expect(patchedOperation).toMatchSnapshot();
      expect([
        ...diffInteractionByOperation(unspecifiedResponse, patchedOperation),
      ]).toHaveLength(0);
    }
  });

  it('generates a patch for a missing response body', () => {
    const operation = operationFixture({
      '200': {
        description: 'success',
        content: {
          'application/json': {},
        },
      },
    });

    const emptyResponse = interactionFixture('200', null);

    let [diff] = [...diffInteractionByOperation(emptyResponse, operation)];
    expect(diff.kind).toBe(OperationDiffResultKind.MissingResponseBody);

    let patches = [...responsesPatches(diff, operation)];
    expect(patches).toHaveLength(1);
    expect(patches).toMatchSnapshot();

    for (let patch of patches) {
      let patchedOperation = OperationPatch.applyTo(patch, operation).expect(
        'operation patch should apply to operation'
      );

      expect(patchedOperation).toMatchSnapshot();
      expect([
        ...diffInteractionByOperation(emptyResponse, patchedOperation),
      ]).toHaveLength(0);
    }
  });
});

function operationFixture(
  responses: OpenAPIV3.ResponsesObject = {}
): Operation {
  return Operation.fromOperationObject('/some-path', HttpMethods.POST, {
    responses,
  });
}

function interactionFixture(
  statusCode: string,
  responseBody: CapturedBody | null
): CapturedInteraction {
  return {
    request: {
      host: 'optic.test',
      method: HttpMethods.POST,
      path: '/some-path',
      body: null,
    },
    response: {
      statusCode,
      body: responseBody,
    },
  };
}
