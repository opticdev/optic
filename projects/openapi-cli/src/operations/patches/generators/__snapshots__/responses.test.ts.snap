// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`responsesPatches generates a patch for a missing response body 1`] = `
Array [
  Object {
    "description": "remove response content for 200",
    "groupedOperations": Array [
      Object {
        "intent": "remove response content (1 bodies)",
        "operations": Array [
          Object {
            "op": "remove",
            "path": "/responses/200/content",
          },
        ],
      },
    ],
    "impact": Array [
      "BackwardsIncompatible",
    ],
  },
]
`;

exports[`responsesPatches generates a patch for a missing response body 2`] = `
Object {
  "method": "post",
  "pathPattern": "/some-path",
  "responses": Object {
    "200": Object {
      "description": "success",
    },
  },
}
`;

exports[`responsesPatches generates a patch for an unspecified response body 1`] = `
Array [
  Object {
    "description": "add text/plain response for 200",
    "groupedOperations": Array [
      Object {
        "intent": "add response body for content type 'text/plain'",
        "operations": Array [
          Object {
            "op": "add",
            "path": "/responses/200/content/text~1plain",
            "value": Object {},
          },
        ],
      },
    ],
    "impact": Array [
      "Addition",
      "BackwardsCompatible",
    ],
  },
]
`;

exports[`responsesPatches generates a patch for an unspecified response body 2`] = `
Object {
  "method": "post",
  "pathPattern": "/some-path",
  "responses": Object {
    "200": Object {
      "content": Object {
        "application/json": Object {},
        "text/plain": Object {},
      },
      "description": "success",
    },
  },
}
`;

exports[`responsesPatches generates a patch for an unspecified response status code 1`] = `
Array [
  Object {
    "description": "add 400 response",
    "groupedOperations": Array [
      Object {
        "intent": "add response status code",
        "operations": Array [
          Object {
            "op": "add",
            "path": "/responses/400",
            "value": Object {
              "description": "automatically documented through Optic",
            },
          },
        ],
      },
    ],
    "impact": Array [
      "Addition",
      "BackwardsCompatible",
    ],
  },
]
`;

exports[`responsesPatches generates a patch for an unspecified response status code 2`] = `
Object {
  "method": "post",
  "pathPattern": "/some-path",
  "responses": Object {
    "200": Object {
      "description": "success",
    },
    "400": Object {
      "description": "automatically documented through Optic",
    },
  },
}
`;
